driver/src/test/functional/com/mongodb/reactivestreams/client/gridfs/GridFSPublisherSpecification.groovy

```diff
@@ -0,0 +1,451 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client.gridfs
+
+import com.mongodb.MongoGridFSException
+import com.mongodb.client.gridfs.model.GridFSFile
+import com.mongodb.client.gridfs.model.GridFSUploadOptions
+import com.mongodb.reactivestreams.client.MongoCollection
+import com.mongodb.reactivestreams.client.MongoDatabase
+import com.mongodb.reactivestreams.client.FunctionalSpecification
+import org.bson.BsonDocument
+import org.bson.BsonString
+import org.bson.Document
+import org.bson.UuidRepresentation
+import org.bson.codecs.UuidCodec
+import spock.lang.Unroll
+
+import java.nio.ByteBuffer
+import java.nio.channels.Channels
+import java.nio.channels.WritableByteChannel
+import java.security.SecureRandom
+
+import static com.mongodb.client.model.Filters.eq
+import static com.mongodb.client.model.Updates.unset
+import static com.mongodb.reactivestreams.client.Fixture.ObservableSubscriber
+import static com.mongodb.reactivestreams.client.Fixture.createPublisher
+import static com.mongodb.reactivestreams.client.Fixture.getDefaultDatabaseName
+import static com.mongodb.reactivestreams.client.Fixture.getMongoClient
+import static com.mongodb.reactivestreams.client.MongoClients.getDefaultCodecRegistry
+import static java.util.concurrent.TimeUnit.SECONDS
+import static org.bson.codecs.configuration.CodecRegistries.fromCodecs
+import static org.bson.codecs.configuration.CodecRegistries.fromRegistries
+
+class GridFSPublisherSpecification extends FunctionalSpecification {
```

Comment:
This is a port of `AsyncGridFSBucketSmokeTestSpecification` with the addition of a subscription cancel test.